plonky3/src/verifier.rs

```diff
@@ -10,250 +12,423 @@ use p3_matrix::dense::RowMajorMatrixView;
+            .values()
+            .map(|opened_values| &opened_values.traces_by_stage[i])
+            .collect_vec()),
+    )
+    .map(|(commit, openings)| {
+        (
+            commit.clone(),
+            trace_domains
+                .iter()
+                .zip_eq(openings)
+                .map(|(trace_domain, StageOpenedValues { local, next })| {
+                    let zeta_next = trace_domain.next_point(zeta).unwrap();
+                    (
+                        *trace_domain,
+                        vec![(zeta, local.clone()), (zeta_next, next.clone())],
+                    )
+                })
                 .collect_vec(),
-            )
-            .chain([(
-                commitments.quotient_chunks.clone(),
-                quotient_chunks_domains
+        )
+    })
+    .collect();
+
+    // for quotient commitments, we have a single commitment, opened on each quotient domain at many points
+    let quotient_chunks_domain_point_and_opens: (_, Vec<(_, _)>) = (
+        proof.commitments.quotient_chunks.clone(),
+        quotient_domains
+            .iter()
+            .zip_eq(opened_values.values())
+            .flat_map(|(domains, values)| {
+                domains
                     .iter()
-                    .zip(&opened_values.quotient_chunks)
-                    .map(|(domain, values)| (*domain, vec![(zeta, values.clone())]))
-                    .collect_vec(),
-            )])
+                    .zip_eq(values.quotient_chunks.iter())
+                    .map(|(domain, chunk)| (*domain, vec![(zeta, chunk.clone())]))
+            })
             .collect_vec(),
-        opening_proof,
-        challenger,
-    )
-    .map_err(VerificationError::InvalidOpeningArgument)?;
+    );
 
-    let zps = quotient_chunks_domains
-        .iter()
-        .enumerate()
-        .map(|(i, domain)| {
-            quotient_chunks_domains
-                .iter()
-                .enumerate()
-                .filter(|(j, _)| *j != i)
-                .map(|(_, other_domain)| {
-                    other_domain.zp_at_point(zeta)
-                        * other_domain.zp_at_point(domain.first_point()).inverse()
-                })
-                .product::<SC::Challenge>()
-        })
-        .collect_vec();
+    let verify_input = preprocessed_domains_points_and_opens
+        .into_iter()
+        .chain(trace_domains_points_and_opens_by_stage)
+        .chain(once(quotient_chunks_domain_point_and_opens))
+        .collect();
 
-    let quotient = opened_values
-        .quotient_chunks
-        .iter()
-        .enumerate()
-        .map(|(ch_i, ch)| {
-            ch.iter()
-                .enumerate()
-                .map(|(e_i, &c)| zps[ch_i] * SC::Challenge::monomial(e_i) * c)
-                .sum::<SC::Challenge>()
-        })
-        .sum::<SC::Challenge>();
+    pcs.verify(verify_input, opening_proof, challenger)
+        .map_err(VerificationError::InvalidOpeningArgument)?;
```

Comment:
do we have tests that fail each of these stages? Maybe not fully needed for this PR, but eventually...

Reply:
Right, we do not, and I think uni-stark doesn't either.